pandas/core/series.py

```diff
@@ -171,17 +171,26 @@ def __init__(self, data=None, index=None, dtype=None, name=None,
                         index = Index(_try_sort(data))
                 try:
                     if isinstance(index, DatetimeIndex):
-                        # coerce back to datetime objects for lookup
-                        data = _dict_compat(data)
-                        data = lib.fast_multiget(data, index.astype('O'),
-                                                 default=np.nan)
+                        if len(data):
+                            # coerce back to datetime objects for lookup
+                            data = _dict_compat(data)
+                            data = lib.fast_multiget(data, index.astype('O'),
+                                                     default=np.nan)
+                        else:
+                            data = np.nan
                     elif isinstance(index, PeriodIndex):
-                        data = [data.get(i, nan) for i in index]
+                        if len(data):
```

Comment:
I would make these 1-liners (and in except
``data = [ data.get(i, np.nan) for i in index ] if len(data) else np.nan``